src/site/content/en/blog/porting-gphoto2-to-the-web/index.md

```diff
@@ -0,0 +1,352 @@
+
+```js
+while (this.canvasRef.current) {
+  try {
+    let blob = await this.props.getPreview();
+
+    let img = await createImageBitmap(blob, { /* … */ });
+    await new Promise(resolve => requestAnimationFrame(resolve));
+    canvasCtx.transferFromImageBitmap(img);
+  } catch (err) {
+    // …
+  }
+}
+```
+
+Using those modern APIs ensures that all of the decoding work is done in the background, and the canvas gets updated only when both the image and the browser are fully prepared for drawing. This achieved consistent 30+ FPS on my laptop, which matched the native performance of both gPhoto2 and the official Sony software.
+
+## Synchronizing the USB access
+
+When a USB data transfer is requested while another operation is already in progress, it will commonly result in a "device is busy" error. Since the preview and the settings UI update regularly, and the user might be trying to capture an image or modify settings at the same time, such conflicts between different operations turned out to be very frequent.
+
+In order to avoid them, I needed to synchronize all the accesses within the application. For that, I've built a promise-based async queue:
+
+```js
+let context = await new Module.Context();
+
+let queue = Promise.resolve();
+
+function schedule(op) {
+  let res = queue.then(() => op(context));
+  queue = res.catch(rethrowIfCritical);
+  return res;
+}
+```
+
+By chaining each operation in a `.then` callback of the existing `queue` promise, and storing the chained result as the new value of `queue`, I can make sure that all operations are executed one by one, in order and without overlaps.
+
+Any operation errors are returned to the caller, while critical (unexpected) errors mark the entire chain as a rejected promise, and ensure that no new operation will be scheduled afterwards.
+
+By keeping the module context in a private (non-exported) variable, I'm minimizing the risks of accessing the `context` by accident somewhere else in the app without going through the `schedule` call.
+
+To tie things together, now each access to the device context has to be wrapped in a `schedule` call like this:
+
+```js
+let config = await this.connection.schedule((context) => context.configToJS());
+```
+
+and
+
+```js
+this.connection.schedule((context) => context.captureImageAsFile());
+```
+
+After that, all the operations were executing successfully without conflicts.
+
+## Conclusion
+
+Together, WebAssembly, Asyncify and Fugu APIs provide a very powerful compilation target for even the most complex applications. They allow you to take a library or an application previously built for a single platform, and port it to the web, making it available to a vastly larger number of users across desktop and mobile devices alike.
```

Comment:
```suggestion
Together, WebAssembly, Asyncify, and Fugu APIs provide a very powerful compilation target for even the most complex applications. They allow you to take a library or an application previously built for a single platform, and port it to the web, making it available to a vastly larger number of users across desktop and mobile devices alike.
```

Reply:
'Powerful' is a badly overused word in web writing. Your readers will likely take this as hyperbole.

Reply:
I'm open to suggestions. Would "capable" work better?

Reply:
If that's your choice. You also have 'compelling', 'impressive', and 'potent'.

Reply:
"capable" seemed best in this context, replaced.